### PR TITLE
[SDK-3290] Add Go API interactive QuickStart

### PR DESCRIPTION
**Pull request by adamjmcgrath**
image

### CODE REVIEW

File: articles/quickstart/backend/golang/interactive.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+useCase: quickstart
+files:
+  - files/jwt
+  - files/main
+---
+
+<!-- markdownlint-disable MD041 MD025 -->
+
+# Add Authorization to a Go application.
+Auth0 allows you to quickly add authentication and gain access to user profile information in your application. This guide demonstrates how to integrate Auth0 with any new or existing ASP.NET Owin Web API application using the [go-jwt-middleware](https://github.com/auth0/go-jwt-middleware) package.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: > any new or existing ASP.NET Owin Web API application

Should this refer to a Go app as opposed to a ASP.NET Owin app?

File: articles/quickstart/backend/golang/interactive.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+contentType: tutorial
+useCase: quickstart
+files:
+  - files/jwt
+  - files/main
+---
+
+<!-- markdownlint-disable MD041 MD025 -->
+
+# Add Authorization to a Go application.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide: titles should use sentence case unless it's referring to an Auth0 product or proper noun

```suggestion
# Add authorization to a Go application.
```

File: articles/quickstart/backend/golang/interactive.md (hunk 1 above)
Comment (stevehobbsdev) [reply articles/quickstart/backend/golang/interactive.md]: > Auth0 allows you to quickly add authentication

Should this be "authentication" or "authorization"? (Title refers to authorization but then we're saying "add authentication").

File: articles/quickstart/backend/golang/interactive.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+```sh
+# The URL of our Auth0 Tenant Domain.
+# If you're using a Custom Domain, be sure to set this to that value instead.
+AUTH0_DOMAIN='${account.namespace}'
+
+# Our Auth0 API's Identifier.
+AUTH0_AUDIENCE='${apiIdentifier}'
+```
+
+## Create a middleware to validate Access Tokens {{{ data-action=code data-code="middleware/jwt.go" }}}
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide (word list): Access Token should be lower case

```suggestion
## Create a middleware to validate access tokens {{{ data-action=code data-code="middleware/jwt.go" }}}
```

File: articles/quickstart/backend/golang/interactive.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+Access Token validation will be done in the `EnsureValidToken` middleware function which can be applied to any
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide (word list): Access Token should be lower case

```suggestion
Access token validation will be done in the `EnsureValidToken` middleware function which can be applied to any
```

File: articles/quickstart/backend/golang/interactive.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+endpoints you wish to protect.
+
+If the token is valid, the resources which are served by the endpoint can be released,
+otherwise a `401 Authorization` error will be returned.
+
+Setup the **go-jwt-middleware** middleware to verify Access Tokens from incoming requests.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide (word list): Access Token should be lower case

```suggestion
Setup the **go-jwt-middleware** middleware to verify access tokens from incoming requests.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+<%= include('../_includes/_api_jwks_description_no_link') %>
+
+In addition to this, include a mechanism for checking that the token has the sufficient **scope** to access the requested resources.
+
+Create a function `HasScope` to check and ensure the Access Token has the correct scope before returning a successful response.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide (word list): Access Token should be lower case

```suggestion
Create a function `HasScope` to check and ensure the access token has the correct scope before returning a successful response.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+## Protect API Endpoints {{{ data-action=code data-code="main.go" }}}
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide: titles should be sentence case

```suggestion
## Protect API endpoints {{{ data-action=code data-code="main.go" }}}
```

File: articles/quickstart/backend/golang/interactive.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+In this example, create an `/api/public` endpoint that does not use the `EnsureToken` middleware as it is accessible to non-authenticated requests.
+
+Create an `/api/private` endpoint that requires the `EnsureToken` middleware as it is only available to authenticated requests containing an Access Token with no additional scope.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide (word list): Access Token should be lower case

```suggestion
Create an `/api/private` endpoint that requires the `EnsureToken` middleware as it is only available to authenticated requests containing an access token with no additional scope.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+Create an `/api/private-scoped` endpoint that requires the `EnsureToken` middleware and `HasScope` as it is only available for authenticated requests containing an Access Token with the `read:messages` scope granted.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide (word list): Access Token should be lower case

```suggestion
Create an `/api/private-scoped` endpoint that requires the `EnsureToken` middleware and `HasScope` as it is only available for authenticated requests containing an access token with the `read:messages` scope granted.
```

**Review by stevehobbsdev — CHANGES_REQUESTED** [review]
Generally looks good, picked up a couple things that conflict with the word list (linked to from the writing style guide).

Also we refer to an ASP.NET Owin web app?

File: articles/quickstart/backend/golang/interactive.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+:::
+
+<%= include('../_includes/_call_api') %>
+
+::::checkpoint
+:::checkpoint-default
+Now that you have configured your application, run your application to verify that:
+* `GET /api/public` is available for non-authenticated requests.
+* `GET /api/private` is available for authenticated requests.
+* `GET /api/private-scoped` is available for authenticated requests containing an Access Token with the `read:messages` scope.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/golang/interactive.md]: Style guide (word list): Access Token should be lower case

```suggestion
* `GET /api/private-scoped` is available for authenticated requests containing an access token with the `read:messages` scope.
```

**Review by stevehobbsdev — CHANGES_REQUESTED** [review]

File: articles/quickstart/backend/golang/interactive.md (hunk 1 above)
Comment (adamjmcgrath) [reply articles/quickstart/backend/golang/interactive.md]: 👍 copy paste error

**Review by stevehobbsdev — DISMISSED** [review]

File: articles/quickstart/backend/golang/interactive.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+* `GET /api/private-scoped` is available for authenticated requests containing an access token with the `read:messages` scope.
+  :::
+
+:::checkpoint-failure
+Sorry about that. Here's a couple things to double check:
+* make sure the token is added as the `Authorization` header
+* does the token have the correct scopes? You can use [jwt.io](https://jwt.io/) to verify.
+
+Still having issues? Check out our [documentation](https://auth0.com/docs) or visit our [community page](https://community.auth0.com) to get more help.
+:::
```
Comment (frederikprijck) [inline articles/quickstart/backend/golang/interactive.md:114]:
```suggestion
:::
::::
```

File: articles/quickstart/backend/golang/interactive.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+<%= include('../_includes/_call_api') %>
+
+::::checkpoint
+:::checkpoint-default
+Now that you have configured your application, run your application to verify that:
+* `GET /api/public` is available for non-authenticated requests.
+* `GET /api/private` is available for authenticated requests.
+* `GET /api/private-scoped` is available for authenticated requests containing an access token with the `read:messages` scope.
+  :::
```
Comment (frederikprijck) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
:::
```

**Review by frederikprijck — CHANGES_REQUESTED** [review]

**Review by frederikprijck — DISMISSED** [review]

**Review by stevehobbsdev — DISMISSED** [review]

File: articles/quickstart/backend/golang/interactive.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  - files/jwt
+  - files/main
+---
+
+<!-- markdownlint-disable MD041 MD025 -->
+
+# Add authorization to a Go application.
+This guide demonstrates how to integrate Auth0 with any new or existing Go API application using the [go-jwt-middleware](https://github.com/auth0/go-jwt-middleware) package.
+
+If you haven't created an API in your Auth0 dashboard yet, you can use the interactive selector to create a new Auth0 API or select an existing API that represents the project you want to integrate with.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
If you have not created an API in your Auth0 dashboard yet, use the interactive selector to create a new Auth0 API or select an existing API for your project.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+Alternatively, you can read [our getting started guide](get-started/auth0-overview/set-up-apis) that helps you set up your first API through the Auth0 dashboard.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
To set up your first API through the Auth0 dashboard, review [our getting started guide](get-started/auth0-overview/set-up-apis).
```

File: articles/quickstart/backend/golang/interactive.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+Every API in Auth0 is configured using an API Identifier that your application code will use as the Audience to validate the access token.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
Each Auth0 API uses the API Identifier, which your application needs to validate the access token.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+<%= include('../../../_includes/_api_auth_intro') %>
+
+## Define permissions
+<%= include('../_includes/_api_scopes_access_resources') %>
+
+## Install dependencies
+
+Add a `go.mod` file to list all the dependencies to be used.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
Add a `go.mod` file to list all the necessary dependencies.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+# If you're using a Custom Domain, be sure to set this to that value instead.
+AUTH0_DOMAIN='${account.namespace}'
+
+# Our Auth0 API's Identifier.
+AUTH0_AUDIENCE='${apiIdentifier}'
+```
+
+## Create a middleware to validate access tokens {{{ data-action=code data-code="middleware/jwt.go" }}}
+
+Access token validation will be done in the `EnsureValidToken` middleware function which can be applied to any
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
The `EnsureValidToken` middleware function validates the access token. You can be apply this function to any endpoints you wish to protect.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+endpoints you wish to protect.
+
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
```

File: articles/quickstart/backend/golang/interactive.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+If the token is valid, the resources which are served by the endpoint can be released,
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
If the token is valid, the endpoint releases the resources. If the token is not valid, the API returns a `401 Authorization` error.
```

File: articles/quickstart/backend/golang/interactive.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+otherwise a `401 Authorization` error will be returned.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion

```

File: articles/quickstart/backend/golang/interactive.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+Setup the **go-jwt-middleware** middleware to verify access tokens from incoming requests.
+
+<%= include('../_includes/_api_jwks_description_no_link') %>
+
+In addition to this, include a mechanism for checking that the token has the sufficient **scope** to access the requested resources.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
Include a mechanism to check that the token has sufficient **scope** to access the requested resources.

File: articles/quickstart/backend/golang/interactive.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+## Protect API endpoints {{{ data-action=code data-code="main.go" }}}
+
+In this example, create an `/api/public` endpoint that does not use the `EnsureToken` middleware as it is accessible to non-authenticated requests.
+
+Create an `/api/private` endpoint that requires the `EnsureToken` middleware as it is only available to authenticated requests containing an access token with no additional scope.
+
+Create an `/api/private-scoped` endpoint that requires the `EnsureToken` middleware and `HasScope` as it is only available for authenticated requests containing an access token with the `read:messages` scope granted.
+
+::: note
+Only the `read:messages` scope is checked. You may want to extend the `HasScope` function or make it a standalone middleware that accepts multiple scopes to fit your use case.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]: checked by what?

File: articles/quickstart/backend/golang/interactive.md [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+:::checkpoint-default
+Now that you have configured your application, run your application to verify that:
+* `GET /api/public` is available for non-authenticated requests.
+* `GET /api/private` is available for authenticated requests.
+* `GET /api/private-scoped` is available for authenticated requests containing an access token with the `read:messages` scope.
+:::
+
+:::checkpoint-failure
+Sorry about that. Here's a couple things to double check:
+* make sure the token is added as the `Authorization` header
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
If your application did not start successfully:
* Verify you added the token as the `Authorization` header
```

File: articles/quickstart/backend/golang/interactive.md [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+* does the token have the correct scopes? You can use [jwt.io](https://jwt.io/) to verify.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/golang/interactive.md]:
```suggestion
* Ensure the token has the correct scopes. Verify with [jwt.io](https://jwt.io/).
```

**Review by tr0njavolta — COMMENTED** [review]
Style guide & some clarity suggestions.

File: articles/quickstart/backend/golang/interactive.md (hunk 22 above)
Comment (stevehobbsdev) [reply articles/quickstart/backend/golang/interactive.md]: @adamjmcgrath could you clarify this? I'm assuming it means that only the `read:messages` scope is validated "by the API"

**Review by tr0njavolta — DISMISSED** [review]
One lingering question on line 95, but approving so I don't block!

File: articles/quickstart/backend/golang/interactive.md (hunk 22 above)
Comment (adamjmcgrath) [reply articles/quickstart/backend/golang/interactive.md]: How about?

```suggestion
Only the `read:messages` scope is checked by the `HasScope` function, you may want to extend it or make it a standalone middleware that accepts multiple scopes to fit your use case.
```

Comment (tr0njavolta) [reply articles/quickstart/backend/golang/interactive.md]:
```suggestion
The `HasScope` function is the only check on the `read:messages` scope. You may want to extend the scope or make a standalone middleware to accept multiple scopes.
```

Still feels kind of awkward but if you think one is clearer, feel free to commit and merge.

**Review by frederikprijck — APPROVED** [review]

**Review by sergiught — APPROVED** [review]

**Pull request merged**